### PR TITLE
[vrotsc, vrotsc-annotations], (#337) Canvas Item For Scheduled Workflow

### LINKED ISSUE #337: Canvas item decorator for Scheduled Workflow
### Description

We need to support Scheduled Workflows with all or close to all of its capabilities.

### Example

This is how the `workflow-item` looks when fetched
```xml
  <workflow-item name="item1" out-name="item0" type="task" launched-workflow-id="9e4503db-cbaa-435a-9fad-144409c08df0">
    <display-name><![CDATA[Other Workflow]]></display-name>
    <script encoded="false"><![CDATA[//Auto generated script, cannot be modified !
var workflowToLaunch = Server.getWorkflowWithId("9e4503db-cbaa-435a-9fad-144409c08df0");
if (workflowToLaunch == null) {
	throw "Workflow not found";
}

var workflowParameters = new Properties();
workflowParameters.put("first",first);
workflowParameters.put("second",second);
scheduledTask = workflowToLaunch.schedule(workflowParameters, workflowScheduleDate);
]]></script>
    <in-binding>
      <bind name="workflowScheduleDate" type="Date" export-name="workflowScheduleDate"/>
      <bind name="first" type="number" export-name="first"/>
      <bind name="second" type="number" export-name="second"/>
    </in-binding>
    <out-binding>
      <bind name="scheduledTask" type="Task"/>
    </out-binding>
    <description><![CDATA[Schedule a workflow and create a task.]]></description>
    <position y="60.0" x="260.0"/>
  </workflow-item>
```

### Notes

- You can add whatever is needed to the decorator to make this work.
- The decorator needs to check if the required input parameters are added, namely the `workflowScheduleDate` (check if it's required)
- We need to support `scheduledTask`. This is pretty much an attribute that will hold information about the scheduled workflow. It's set by the item itself, not by us and can be empty. image
- Everything else should work as normal

### PR DESCRIPTION
**Pull request by Michaelpalacce**
#### `@ScheduledWorkflowItem`

The decorator is used to specify a scheduled workflow item that will be called.

##### Supported Parameters

- `target` - The name of the next in line item. Same as `@Item`.
- `linkedItem` - The ID of the workflow to schedule.

In order to bind inputs and outputs, you do it with the `@In` and `@Out` decorators. This is the same way we do it for other items.

##### Inputs

Special input is needed for the ScheduledWorkflowItem.

- `workflowScheduleDate` - {Date} is required. The name **must** be `workflowScheduleDate`. If this is missing an error is thrown. We don't check if the type is `Date` but Aria Orchestrator will complain.

##### Outputs

Special output is needed for the ScheduledWorkflowItem.

- `scheduledTask` - {Task} is optional. If it's missing nothing will happen, if it's added, then the name **must** be `scheduledTask`. This is the task that is scheduled.

```typescript
import { Workflow, Out, In, Item, RootItem, DecisionItem, WaitingTimerItem, WorkflowItem, ScheduledWorkflowItem } from "vrotsc-annotations";

@Workflow({
	name: "Example Waiting Timer",
	path: "VMware/PSCoE",
	attributes: {
		waitingTimer: {
			type: "Date"
		},
		counter: {
			type: "number"
		},
		first: {
			type: "number"
		},
		second: {
			type: "number"
		},
		result: {
			type: "number"
		},
		workflowScheduleDate: {
			type: "Date"
		},
		scheduledTask: {
			type: "Task"
		}
	}
})
export class HandleNetworkConfigurationBackup {
	@DecisionItem({ target: "waitForEvent", else: "prepareItems" })
	public decisionElement(waitingTimer: Date) {
		return waitingTimer !== null;
	}

	@Item({ target: "callOtherWf" })
	public prepareItems(@In @Out first: number, @In @Out second: number, @In @Out workflowScheduleDate: Date) {
		first = 1;
		second = 2;
		workflowScheduleDate = System.getDate("1 minute from now", undefined);
	}

	@WorkflowItem({
		target: "print",
		linkedItem: "9e4503db-cbaa-435a-9fad-144409c08df0"
	})
	public callOtherWf(@In first: number, @In second: number, @Out result: number) {
	}


	@Item({ target: "scheduleOtherWf" })
	public print(@In result: number) {
		System.log("Result: " + result);
	}

	@ScheduledWorkflowItem({
		target: "printScheduledDetails",
		linkedItem: "9e4503db-cbaa-435a-9fad-144409c08df0"
	})
	public scheduleOtherWf(@In first: number, @In second: number, @In workflowScheduleDate: Date, @Out scheduledTask: Task) {
	}

	@Item({ target: "end" })
	public printScheduledDetails(@In scheduledTask: Task) {
		System.log(`Scheduled task: ${scheduledTask.id}, [${scheduledTask.state}]`);
	}

	@Item({ target: "decisionElement", exception: "" })
	public execute(@Out @In waitingTimer: Date, @Out @In counter: number): void {
		if (!counter) {
			counter = 0;
		}

		counter++;

		if (counter < 2) {
			const tt = Date.now() + 5 * 1000;
			waitingTimer = new Date(tt);
		} else {
			waitingTimer = null;
		}

		System.log("Counter: " + counter);
		System.log("Waiting Timer: " + waitingTimer);
	}

	@Item({ target: "execute", exception: "" })
	@RootItem()
	public start() {
		System.log("Starting workflow");
	}

	@WaitingTimerItem({ target: "execute" })
	public waitForEvent(@In waitingTimer: Date) {
	}
}
```
Will give the following:

image

image

image

### Checklist

- [x] I have added relevant error handling and logging messages to help troubleshooting
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation, relevant usage information (if applicable)
- [x] I have updated the PR title with affected component, related issue number and a short summary of the changes introduced
- [x] I have added labels for implementation kind (kind/*) and version type (version/*)
- [x] I have tested against live environment, if applicable
- [x] I have synced any structure and/or content vRA-NG improvements with vra-ng and ts-vra-ng archetypes (if applicable)
- [x] I have my changes rebased and squashed to the minimal number of relevant commits. **Notice: don't squash all commits**
- [x] I have added a descriptive commit message with a short title, including a `Fixed #XXX -` or `Closed #XXX -` prefix to auto-close the issue

### Testing

e2e tests added to the pr and screenshots as proof it works

Fixed #337

### CODE REVIEW

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/helpers/sourceFile.ts [hunk 1]
```diff
@@ -12,19 +12,20 @@
...
-	printSourceFile(methodNode: MethodDeclaration, sourceFile: SourceFile): string;
+	printSourceFile(methodNode: MethodDeclaration, sourceFile: SourceFile, itemInfo?: WorkflowItemDescriptor): string;
 }
 
 /**
  * Default source file printer will directly print the source file with the method node body statements.
  */
 export class DefaultSourceFilePrinter implements SourceFilePrinter {
-	public printSourceFile(methodNode: MethodDeclaration, sourceFile: SourceFile): string {
+	public printSourceFile(methodNode: MethodDeclaration, sourceFile: SourceFile, itemInfo: WorkflowItemDescriptor): string {
```
Comment (akantchev) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/helpers/sourceFile.ts:28]: The `itemInfo` parameter should be optional (as in the interface is optional.

Comment (Michaelpalacce) [reply typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/helpers/sourceFile.ts:28]: In the interface it should not be optional, will fix it

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/scheduledWorkflowItemDecoratorStrategy.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	  <bind name="scheduledTask" type="Task"/>
+	</out-binding>
+	<description><![CDATA[Schedule a workflow and create a task.]]></description>
+	<position y="60.0" x="260.0"/>
+  </workflow-item>
+ * ```
+ */
+export default class ScheduledWorkflowItemDecoratorStrategy implements CanvasItemDecoratorStrategy {
+	constructor(private readonly sourceFilePrinter: SourceFilePrinter = new ScheduledWorkflowItemSourceFilePrinter()) { }
+
```
Comment (akantchev) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/scheduledWorkflowItemDecoratorStrategy.ts:57]: It is a good idea to add JSDoc to each of the public methods.

File: docs/versions/latest/Release.md [hunk 3]
```diff
@@ -140,7 +140,151 @@ export class WorkflowEnd {
...
+  public printScheduledDetails(@In scheduledTask: Task) {
+    System.log(`Scheduled task: ${scheduledTask.id}, [${scheduledTask.state}]`);
+  }
+
+
+  @Item({ target: "prepareItems", exception: "" })
+  @RootItem()
+  public start() {
+    System.log("Starting workflow");
+=======
```
Comment (akantchev) [inline docs/versions/latest/Release.md:220]: Please remove (probably leftover from the merge conflict).

File: typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/scheduledWorkflowItemDecoratorStrategy.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	registerItemArguments(itemInfo: WorkflowItemDescriptor, decoratorNode: Decorator): void {
+		getDecoratorProps(decoratorNode).forEach((propTuple) => {
+			const [propName, propValue] = propTuple;
+			switch (propName) {
+				case "target":
+					itemInfo.target = propValue;
+					break;
+
+				case "exception":
```
Comment (akantchev) [inline typescript/vrotsc/src/compiler/transformer/fileTransformers/workflow/decorators/scheduledWorkflowItemDecoratorStrategy.ts:74]: I see you didn't declare the `exception` parameter in the `vro-types/vrotsc-annotations/index.d.ts` nor you're handling it in the decorator. You should remove this case if you don't need to handle the `exception` parameter.

**Review by akantchev — APPROVED** [review]

**Review by akantchev — APPROVED** [review]

**Review by akantchev — APPROVED** [review]

**Pull request merged**